tests/jax/components/training/trainer_test.py

```diff
@@ -0,0 +1,680 @@
+    """Creates mock trainer
+
+    Doesn't have shared weights and has fixed agent
+    trainer sampling setup.
+    """
+
+    mock_trainer = Trainer(store=SimpleNamespace(), components=[])
+
+    mock_trainer.store.agents = ["agent_0", "agent_1", "agent_2"]
+    mock_trainer.store.trainer_id = "trainer_0"
+    mock_trainer.store.table_network_config = {
+        "trainer_0": ["network_agent", "network_agent", "network_agent"]
+    }
+
+    return mock_trainer
+
+
+@pytest.fixture
+def mock_one_trainer_per_network_no_shared_weights_fixed_sampling() -> Trainer:
+    """Creates mock trainer
+
+    Doesn't have shared weights and has fixed agent
+    trainer sampling setup.
+    """
+
+    mock_trainer = Trainer(store=SimpleNamespace(), components=[])
+
+    mock_trainer.store.agents = ["agent_0", "agent_1", "agent_2"]
+    # trainer id must be specified per test
+    mock_trainer.store.table_network_config = {
+        "trainer_0": ["network_agent_0", "network_agent_1", "network_agent_2"],
+        "trainer_1": ["network_agent_0", "network_agent_1", "network_agent_2"],
+        "trainer_2": ["network_agent_0", "network_agent_1", "network_agent_2"],
+    }
+
+    return mock_trainer
+
+
+@pytest.fixture
+def mock_one_trainer_per_network_random_sampling() -> Trainer:
+    """Creates mock trainer
+
+    Doesn't have shared weights and has random agent
+    trainer sampling setup.
+    """
+
+    mock_trainer = Trainer(store=SimpleNamespace(), components=[])
+
+    # trainer id must be given at each test
+    mock_trainer.store.agents = ["agent_0", "agent_1", "agent_2"]
+    mock_trainer.store.table_network_config = {
+        "trainer_0": ["network_0"],
+        "trainer_1": ["network_1"],
+        "trainer_2": ["network_2"],
+    }
+
+    return mock_trainer
+
+
+############################
+# TESTS
+############################
+
+# ON_BUILDING_INIT_END TESTS
+
+
+def test_single_trainer_shared_weights_fixed_sampling(
+    mock_builder_shared_weights_fixed_sampling: Builder,
+) -> None:
+    """Tests on_building_init_end hook.
+
+    Single trainer for all networks, shared network weights and
+    fixed agent network sampling.
+    """
+
+    mock_trainer = SingleTrainerInit()
+
+    builder = mock_builder_shared_weights_fixed_sampling
+    mock_trainer.on_building_init_end(builder)
+
+    assert builder.store.net_spec_keys == {"network_agent": "agent_0"}
+    assert builder.store.table_network_config == {
+        "trainer": ["network_agent", "network_agent", "network_agent"]
+    }
+    assert builder.store.trainer_networks == {"trainer": ["network_agent"]}
+    assert builder.store.networks == "network_initialized"
+
+
+def test_single_trainer_no_shared_weights_fixed_sampling(
+    mock_builder_no_shared_weights_fixed_sampling: Builder,
+) -> None:
+    """Tests on_building_init_end hook.
+
+    Single trainer for all networks, no shared network weights and
+    fixed agent network sampling.
+    """
+
+    mock_trainer = SingleTrainerInit()
```

Comment:
This could be a fixture, see elsewhere too. But up to you. I kinda back how it is at the moment, given the number of fixtures already.

Reply:
Made them into fixtures :call_me_hand: